misc/experimental/promise/api/celix/Promise.h

```diff
@@ -0,0 +1,512 @@
+         * <p/>
+         * If this Promise is successfully resolved, the returned Promise will be resolved with the value of specified
+         * Function as applied to the value of this Promise. If the specified Function throws an exception, the returned
+         * Promise will be failed with the exception.
+         * <p/>
+         * If this Promise is resolved with a failure, the returned Promise will be failed with that failure.
+         * <p/>
+         * This method may be called at any time including before and after this Promise has been resolved.
+         *
+         * @tparam R     The value type associated with the returned Promise.
+         * @param mapper The Function that will map the value of this Promise to the value that will be used to resolve the
+         *               returned Promise. Must not be null.
+         * @return A Promise that returns the value of this Promise as mapped by the specified Function.
+         */
+        template<typename R>
+        celix::Promise<R> map(std::function<R(T)> mapper);
+
+        /**
+         * Filter the value of this Promise.
+         * <p/>
+         * If this Promise is successfully resolved, the returned Promise will either be resolved with the value of this
+         * Promise if the specified Predicate accepts that value or failed with a NoSuchElementException if the specified
+         * Predicate does not accept that value. If the specified Predicate throws an exception, the returned Promise will
+         * be failed with the exception.
+         * <p/>
+         * If this Promise is resolved with a failure, the returned Promise will be failed with that failure.
+         * <p/>
+         * This method may be called at any time including before and after this Promise has been resolved.
+         *
+         * @param predicate The Predicate to evaluate the value of this Promise.
+         * @return A Promise that filters the value of this Promise.
+         */
+        Promise<T> filter(std::function<bool(T)> predicate);
+
+        /**
+         * Time out the resolution of this Promise.
+         * <p>
+         * If this Promise is successfully resolved before the timeout, the returned
+         * Promise is resolved with the value of this Promise. If this Promise is
+         * resolved with a failure before the timeout, the returned Promise is
+         * resolved with the failure of this Promise. If the timeout is reached
+         * before this Promise is resolved, the returned Promise is failed with a
+         * {@link TimeoutException}.
+         *
+         * @param duration  The time to wait. Zero and negative
+         *            time is treated as an immediate timeout.
+         * @return A Promise that is resolved when either this Promise is resolved
+         *         or the specified timeout is reached.
+         */
+        template<typename Rep, typename Period>
+        Promise<T> timeout(std::chrono::duration<Rep, Period> duration);
```

Comment:
I was wondering if a timeout on the getValue made sense, but this is a better solution.